### PR TITLE
feat: admin http api

### PR DESCRIPTION
**Pull request by jun0315**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

support admin meta http api.

### feature
1. Query the catalog list. 
2. Query the schema list by catalog name. 
3. Query the table list by catalog name and schema name.
4. Query table info by full table name.
5. Query the information of the meta leader node.
6. Query heartbeat by ip.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta-srv/src/service/admin/meta.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        match map.get("schema_name") {
+            Some(schema_value) => key_prefix = key_prefix + schema_value + SEPARATOR,
+            None => {
+                return error::MissingRequiredParameterSnafu {
+                    msg: "schema_name parameter is required".to_string(),
+                }
+                .fail();
+            }
+        }
+        get_key_list_response_by_prefix_key(&key_prefix, &self.kv_store).await
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/meta.rs:103]:
```suggestion
        let catalog_name = match map.get("catalog_name") {
            Some(catalog_name) => catalog_name,
            None => {
                return error::MissingRequiredParameterSnafu {
                    msg: "catalog_name parameter is required".to_string(),
                }
                .fail();
            }
        };

        let schema_name = match map.get("schema_name") {
            // Some(schema_value) => key_prefix = key_prefix + schema_value + SEPARATOR,
            Some(schema_name) => schema_name,
            None => {
                return error::MissingRequiredParameterSnafu {
                    msg: "schema_name parameter is required".to_string(),
                }
                .fail();
            }
        };

        let prefix = format!("{}-{}-{}", TABLE_KEY_PREFIX, catalog_name, schema_name);
        get_key_list_response_by_prefix_key(&prefix, &self.kv_store).await

```

File: src/meta-srv/src/service/admin/meta.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        http::Response::builder()
+            .status(http::StatusCode::OK)
+            .body(value)
+            .context(error::InvalidHttpBodySnafu)
+    }
+}
+
+/// Get kv_store's key list with http response format by prefix key
+async fn get_key_list_response_by_prefix_key(
+    key_prefix: &String,
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/meta.rs:142]: Avoid clone of String
```suggestion
    key_prefix: String,
```

**Comment by Fengys123** [comment]
Suggest add some unit tests.

File: src/meta-srv/src/service/admin/heartbeat.rs [hunk 3]
```diff
@@ -29,14 +29,28 @@ pub struct HeartBeatHandler {
 
 #[async_trait::async_trait]
 impl HttpHandler for HeartBeatHandler {
-    async fn handle(&self, _: &str, _: &HashMap<String, String>) -> Result<http::Response<String>> {
+    async fn handle(
+        &self,
+        _: &str,
+        map: &HashMap<String, String>,
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/heartbeat.rs:35]:
```suggestion
        params: &HashMap<String, String>,
```

File: src/meta-srv/src/service/admin/heartbeat.rs [hunk 4]
```diff
@@ -29,14 +29,28 @@ pub struct HeartBeatHandler {
...
-        let stat_vals: Vec<StatValue> = stat_kvs.into_values().collect();
+        let mut stat_vals: Vec<StatValue> = stat_kvs.into_values().collect();
+
+        if let Some(addr) = map.get("addr") {
+            stat_vals.retain(|stat_val| {
+                if let Some(stat) = stat_val.stats.get(0) {
+                    stat.addr == addr.clone()
+                } else {
+                    false
+                }
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/heartbeat.rs:51]: It seems that clone can be removed to reduce one memory copy. 

And more elegant code as follows? 
```suggestion
                stat_val
                    .stats
                    .get(0)
                    .map(|stat| &stat.addr == addr)
                    .unwrap_or(false)

```

File: src/meta-srv/src/service/admin/leader.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            http::Response::builder()
+                .status(http::StatusCode::OK)
+                .body(leader_addr)
+                .context(error::InvalidHttpBodySnafu)
+        } else {
+            http::Response::builder()
+                .status(http::StatusCode::OK)
+                .body("election info is None".to_string())
+                .context(error::InvalidHttpBodySnafu)
+        }
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/leader.rs:41]:
```suggestion
        if let Some(election) = &self.election {
            let leader_addr = election.leader().await?.0;
            return http::Response::builder()
                .status(http::StatusCode::OK)
                .body(leader_addr)
                .context(error::InvalidHttpBodySnafu);
        }

        http::Response::builder()
            .status(http::StatusCode::OK)
            .body("election info is None".to_string())
            .context(error::InvalidHttpBodySnafu)

```

File: src/meta-srv/src/service/admin/meta.rs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    ) -> Result<http::Response<String>> {
+        let catalog = match params.get("catalog_name") {
+            Some(catalog) => catalog,
+            None => {
+                return error::MissingRequiredParameterSnafu {
+                    param: "catalog_name",
+                }
+                .fail();
+            }
+        };
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/meta.rs:66]:
```suggestion
        let catalog = params.get("catalog_name").context(error::MissingRequiredParameterSnafu {
                    msg: "catalog_name parameter is required".to_string(),
        })?;

```

File: src/meta-srv/src/service/admin/meta.rs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+        let schema = match params.get("schema_name") {
+            Some(schema) => schema,
+            None => {
+                return error::MissingRequiredParameterSnafu {
+                    param: "schema_name",
+                }
+                .fail();
+            }
+        };
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/meta.rs:97]: Same as above. please use `context()`.

File: src/meta-srv/src/service/admin/meta.rs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    ) -> Result<http::Response<String>> {
+        let table_name = match params.get("full_table_name") {
+            Some(full_table_name) => full_table_name.replace('.', "-"),
+            None => {
+                return error::MissingRequiredParameterSnafu {
+                    param: "full_table_name",
+                }
+                .fail();
+            }
+        };
```
Comment (Fengys123) [inline src/meta-srv/src/service/admin/meta.rs:118]: Same as above. please use context().

File: src/meta-srv/src/service/admin/meta.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+    pub kv_store: KvStoreRef,
+}
+
+pub struct TablesHandler {
+    pub kv_store: KvStoreRef,
+}
+
+pub struct TableHandler {
+    pub kv_store: KvStoreRef,
+}
```
Comment (shuiyisong) [inline src/meta-srv/src/service/admin/meta.rs:42]: 4 different but similar structs with same inner value, just to use `HttpHandler`. Seems too tedious. 

Can we simplify that by change `HttpHandler` or use http path as input param?

File: src/meta-srv/src/service/admin/meta.rs (hunk 8 above)
Comment (jun0315) [reply src/meta-srv/src/service/admin/meta.rs:118]: It's elegant to write like this!

File: src/meta-srv/src/service/admin/meta.rs (hunk 9 above)
Comment (jun0315) [reply src/meta-srv/src/service/admin/meta.rs:42]: In the current http framework of admin, the url of our route corresponds to a handler.  https://github.com/evenyag/greptimedb/blob/f8508bb6e7ffb5f409e335134b1473b1a2c1019e/src/meta-srv/src/service/admin.rs#L46  A handler needs to implement handle fn. In addition to the current writing method, I can't think of any other way at present. 0.0

**Review by Fengys123 — APPROVED** [review]
LGTM

**Review by shuiyisong — APPROVED** [review]

**Pull request merged**